### ISSUE TITLE
Format for MathML (and maybe others) on the clipboard is wrong

### ISSUE THREAD

**Issue by NSoiffer** [arnog/mathlive#1496]
## Description

According to the documentation at https://cortexjs.io/mathlive/guides/interacting/, copying an expression should put out a format usable by MathML-aware programs when pasted into them. Using `application/mathml+xml` isn't correct for Windows. That's the media type, not a clipboard format "flavor". I don't believe it is correct for Macs either. From the [MathML spec](https://www.w3.org/TR/MathML3/chapter6.html#encoding-names):
* Windows Clipboard Flavor: MathML
* Universal Type Identifier: public.mathml

MS Word, Mathematica, and Maple are all MathML clipboard-aware, so copying into them is a way to test. Copying out of them is also a way to see the proper format on the clipboard.

I apologize for the weak bug report, but I'm in the middle of fixing something else and was afraid I would forget to file something if I didn't write something up now.

## Notes:
1. What is placed on the clipboard should begin with the processing instruction `<?xml version="1.0"?>`.
2. On Windows (the only place I've done an implementation), you need register the MathML "flavor" -- it is not built-in. If it is of some help, some python code I wrote does something like:
```
        winUser.emptyClipboard()
        windll.user32.RegisterClipboardFormatW("MathML")

        # 'text' is MathML prepended with '<?xml...>
        let bufLen = len(text.encode(WCHAR_ENCODING, errors="surrogatepass")) + 2 '
        let h=winKernel.HGLOBAL.alloc(winKernel.GMEM_MOVEABLE, bufLen)
        # Acquire a lock to the global memory receiving a local memory address
        with h.lock() as addr:
            # Write the text into the allocated memory
            let buf=(c_wchar*bufLen).from_address(addr)
            buf.value=text

       # Set the clipboard data with the global memory
       if not windll.user32.SetClipboardData(format,h):
         raise ctypes.WinError()

       # NULL the global memory handle so that it is not freed at the end of scope as the clipboard now has it.
       h.forget()
```
3. I did not check, but I suspect the other specialized formats on the clipboard might be problematic.

**Comment by arnog** [comment]
According to the HTML spec, the format string should be `A Unicode string giving the type or format of the data, generally given by a [MIME type](https://mimesniff.spec.whatwg.org/#mime-type). Some values that are not [MIME types](https://mimesniff.spec.whatwg.org/#mime-type) are special-cased for legacy reasons.`.

As I understand it, that string is not intended to be a native format type, but the responsibility to "translate" it if necessary to something the host operating system  can handle is left to the UA.

Having data types that are platform-specific seems difficult to make workable for a web app. What format should be used for ChromeOS or for Linux?

Admittedly, a format type that no one else is going to expect to read from is not very useful either.

It would be great if there could be some standardization effort to decide on the name of that format, across platforms, include the web platform, and if whatever interaction with the host operating system to support this type, such as making calls to system APIs which are not available from the web platform (as you indicate in your python sample), would be done by the host browsers.

**Comment by polx** [comment]
Hey @arnog and @NSoiffer , 

I think we're confusing the world outside the browser and inside the browser.
[DataTransfer's API](https://html.spec.whatwg.org/multipage/dnd.html#the-datatransfer-interface) should expose content by media-types while the host OS does generally not do that (mac, ios, android and windows don't).

My experience, however, is that the media-types beyond the most primitive ones (text/images/urls) are not transferred outside of the browser. So MathML is not thus far. I think that clipboard pickling might solve this one day.

Are you @NSoiffer reporting about MathLive within a browser? Which browser? The bit of code from Windows is for a native app or a privileged extension, right?

Are you @arnog saying that you expose this outside? At least on firefox or chromium on the mac today, selecting a bit of math on the [demo](https://cortexjs.io/mathlive/) page, what's put in the clipboard is plain text TeX and a kind of proprietary format. The [documentation](https://cortexjs.io/mathlive/guides/interacting/) is not clear about the bounds between outer OS world and browser-world. 

Maybe I should reopen the issue but I don't think I'm allowed.

**Comment by arnog** [comment]
@polx You are correct: there are two different ways to indicate the format of the flavors contained in a clipboard:
- in the context of the web platform, i.e. using the JavaScript APIs available within a browser, the format is specified with the `clipboardData.setData()` API. The HTML spec recommends it to be a MIME-type or other string.
- in the context of a native app, the format on the clipboard is defined by the host operating system clipboard API. In the case of macOS, it's typically a UTI (a reverse dot notation string) and in the case of Windows it's a numeric identifier (with the possibility to register additional non-standard identifiers, using the sample code that Neil shared above).

The browsers (UA) have  latitude on how to translate the `clipboardData.setData()` call and its MIME-type-like format argument to a format understood by the host operating system. Browsers typically do this for a few standard formats such as text, HTML, images, and lists of items.

When using a custom format which is not recognized/handled by the hosts browsers, all three browsers I tested (on macOS), store the data using a custom, browser specific, flavor that "wraps" the content of the custom data:

- `com.apple.WebKit.custom-pasteboard-data` on Safari
- `org.mozilla.custom-clipdata` on Firefox
- `org.chromium.web-custom-data` on Chrome. 

These flavors are not interoperable with native apps, but they are interoperable with other web apps on the same browser (i.e. other web apps can get access to the custom flavor stored in the clipboard). 

In the case of MathLive, the custom flavors it uses, such as `application/json` can be used to copy and paste between mathfields on the same page (or on different pages, for that matter). But this custom data is not directly available to native apps (unless they specifically look for the browser specific native flavors, and decode the payload, which seems to be a different format for all three browsers).

MathLive used to store a custom flavor for MathML on the clipboard, using a MIME-type, as per the conventions of the JavaScript `clipboardData.setData()` API. 

I believe that Neil pointed out that using this MIME-type is not the standard flavor for MathML content. That's a fair point. Since it's not standard, and since the browsers don't recognize it and handle it, it is not exposed to the native apps either. So, it is "locked" inside the web platform world. But since no-one seems to make use of it, I have removed it. So, MathLive no longer puts a non-standard MathML content on the clipboard.

Since @NSoiffer is involved in the standardization effort of MathML, I would suggest that the following would be desirable for the standard group to consider:
- in addition to defining OS specific clipboard flavor formats, to also define a web platform clipboard flavor. I would suggest using the MIME-type for MathML, since using a MIME-type is already what the HTML spec recommends for the web platform clipboard API
- encourage the browsers that support MathML to recognize when the `clipboardData.setData()` API is used with the web platform clipboard flavor for MathML and when that is the case that they "translate" it (i.e. put as a separate flavor) to the host operating system system clipboard, using the OS specific flavor defined by the MathML spec. 
- conversely, when the system clipboard contains a MathML flavor from a native application, it should get converted by the web browsers supporting MathML to a "web platform flavor", i.e. such that `clipboardData.getData()` using the MathML web platform flavor type (presumably the MIME-type for MathML) would be able to be retrieved by web apps.

Once the MathML spec has been updated and these changes deployed in some browsers, it will then make sense for MathLive to add support for a standard MathML content on the clipboard.
